primitives/npos-elections/src/lib.rs

```diff
@@ -144,9 +145,84 @@ pub type VoteWeight = u64;
 /// A type in which performing operations on vote weights are safe.
 pub type ExtendedBalance = u128;
 
-/// The score of an assignment. This can be computed from the support map via
-/// [`EvaluateSupport::evaluate`].
-pub type ElectionScore = [ExtendedBalance; 3];
+/// The score of an election. This is the main measure of an election's quality.
+#[derive(Clone, Copy, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo, Debug, Default)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct ElectionScore {
+	/// The minimal winner, in terms of total backing stake.
+	///
+	/// This parameter should be maximized.
+	pub minimal_stake: ExtendedBalance,
+	/// The sum of the total backing of all winners.
+	///
+	/// This parameter should maximized
+	pub sum_stake: ExtendedBalance,
+	/// The sum squared of the total backing of all winners, aka. the variance.
+	///
+	/// Ths parameter should be minimized.
+	pub sum_stake_squared: ExtendedBalance,
+}
+
+impl ElectionScore {
+	fn iter(&self) -> impl Iterator<Item = ExtendedBalance> {
+		vec![self.minimal_stake, self.sum_stake, self.sum_stake_squared].into_iter()
+	}
```

Comment:
`iter` function should be remove, iterator items should have same semantic. Would maybe make part of other code less good looking, but more straightforward.

Reply:
How about if I rename it to `iter_by_importance`? that would clearly imply that you are getting an iterator, in which the items are returned from the most important to the least, which is exactly what you are getting. 

The main problem I have with your suggestion is that `vec![self.minimal_stake, self.sum_stake, self.sum_stake_squared]` is an error-prone code, so I don't want to repeat it more than once.

Reply:
7abd0abba1637de7598103b784b9e2fc7b228a6e

Reply:
suggestion was more to remove all array of score and only use the new struct.
Codes for comparing will not use iter anymore, probably will be less compact, but would be very easy to read (currently we have to check the iter function to understand which index correspond to which field),  and there is a Vec involve.

Reply:
this function is only used in one place, and, again, my main goal is to avoid code duplication. How about if I move it to a closure inside the function where it is used?

Reply:
Yes, something along the line of :
```
fn compare(this: Self, that: Self) -> (bool, Ordering) {
  (this.ge(that), this.tcmp(that, threshold)
}
match [
  compare(this.threshold, that.threshold),
  compare(this.minimal_stake, that.minimal_stake),
  compare(this.sum_stake_squared, that.sum_stake_squared),
] {
```
can be a bit more verbose but avoid two vec allocations.

Reply:
yeah allocations are pretty insignificant here, will move to closure.

Reply:
okay closure is a hassle as well since you can't return `impl Trait`. I don't see any difference between that an a private function called `iter_by_significance`. hope 8d28d210470b811e8b52e7663f78bb5ed15d38d2 is good enough.